tools/indexer/cmd/indexer/main.go

```diff
@@ -0,0 +1,434 @@
+
+	// Setup graceful shutdown
+	ctx, cancel := signal.NotifyContext(context.Background(), syscall.SIGINT, syscall.SIGTERM)
+	defer cancel()
+
+	// Connect to StarRocks database
+	db, err := database.MustConnect(ctx, dbURL, 20, 5)
+	if err != nil {
+		log.Fatalf("[DB] Connection failed: %v", err)
+	}
+	defer db.Close()
+	log.Printf("[DB] Connected to StarRocks database")
+
+	// Ensure required tables exist
+	if err := db.EnsureStateTable(ctx); err != nil {
+		log.Fatalf("[DB] Failed to ensure state table: %v", err)
+	}
+	log.Printf(" [DB] State table ready")
+
+	// Initialize HTTP client
+	httpc := httputil.NewHTTPClient(c.Duration("http-timeout"))
+	log.Printf("[HTTP] Client initialized with %s timeout", c.Duration("http-timeout"))
+
+	// Load relay configurations
+	relays, err := relay.UpsertRelaysAndLoad(ctx, db)
+	if err != nil {
+		log.Fatalf("[RELAYS] Failed to load: %v", err)
+	}
+	log.Printf("[RELAYS] Loaded %d active relays:", len(relays))
+	for _, r := range relays {
+		log.Printf("  - Relay ID %d: %s", r.ID, r.URL)
+	}
+
+	// Initialize starting block number
+	lastBN, found := db.LoadLastBlockNumber(ctx)
+	if !found || lastBN == 0 {
+		log.Printf(" [INIT] No previous state found, checking database for latest block...")
+		err := db.Conn.QueryRowContext(ctx, `SELECT COALESCE(MAX(block_number),0) FROM blocks`).Scan(&lastBN)
+		if err != nil {
+			log.Printf("[INIT] Database query failed: %v", err)
+		}
+	}
+
+	// Replace the hardcoded block search with:
+	if lastBN == 0 {
+		log.Printf("[INIT] Getting latest block from Ethereum RPC...")
+
+		latestBlock, err := ethereum.GetLatestBlockNumber(httpc, infuraRPC)
+		if err != nil {
+			log.Fatalf("[INIT] Failed to get latest block from RPC: %v", err)
+		}
+
+		lastBN = latestBlock - 10 // Start 10 blocks behind to ensure data availability
+		log.Printf("[INIT] Starting from block %d (latest: %d)", lastBN, latestBlock)
+	}
+
+	log.Printf(" [INIT] Starting from block number: %d", lastBN)
+	log.Printf("[INIT] Indexer configuration - Lookback: %d slots, Batch size: %d",
+		c.Int("backfill-lookback-slots"), c.Int("backfill-batch"))
+
+	// Setup tickers
+	backfillTicker := time.NewTicker(c.Duration("backfill-batch"))
+	defer backfillTicker.Stop()
+
+	mainTicker := time.NewTicker(c.Duration("block-interval"))
+	defer mainTicker.Stop()
+
+	log.Printf("🎉 [INIT] Blockchain indexer started successfully")
+
+	// Main processing loop
+	for {
+		select {
+		case <-ctx.Done():
+			log.Printf(" [SHUTDOWN] Graceful shutdown initiated: %v", ctx.Err())
+			if err := db.SaveLastBlockNumber(ctx, lastBN); err != nil {
+				log.Printf("[SHUTDOWN] Failed to save last block number: %v", err)
+			}
+			log.Printf("[SHUTDOWN] Indexer stopped at block %d", lastBN)
+			return nil
+
+		case <-backfillTicker.C:
+			log.Printf("[BACKFILL] Starting backfill operations...")
+			backfill.RunAll(ctx, db, httpc, createOptionsFromCLI(c), relays)
```

Comment:
Why is ticker needed for backfill? This operation should only happen once. Also it can run in parallel with the main go routine so it should just be started before we enter this loop.